### PR TITLE
Sdk automation script

### PR DESCRIPTION
**Pull request by ShivangiReja**
SDK automation task basically involves creating a PR in azure-sdk-for-net repo when something changes in the swagger files in azure-rest-api-specs repo. This will result in removing manual work for updating the SDKs in azure-sdk-for-net repo when something changes in the swagger files. 

SDK Automation Doc: https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/sdkautomation/sdk_customization.md

This PR is for Implementing powershell script which takes[ json file](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/sdkautomation/sdk_customization.md#generateinput-example) as an input, updates the input-file path in autorest.mds for the changed swagger files, generates sdk using track-2 code gen and produces output json as mentioned [here](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/sdkautomation/sdk_customization.md#generateoutput-example). 

For reference here are the examples of input and output json in the test env:
* [generateInput.json](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=107)
* [generateOutput.json](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=271)

### CODE REVIEW

File: eng/scripts/Sdk-Generate.ps1 [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+Write-Host "List of changed swagger files and related readmes:`n$changedFilePaths`n"
+
+$autorestFilesPath = Get-ChildItem -Path "$RepoRoot/sdk"  -Filter autorest.md -Recurse | Resolve-Path -Relative
+
+Write-Host "Updating autorest.md files for all the changed swaggers."
+$sdksInfo = @{}
+$headSha = $inputFileContent.headSha
+$repoHttpsUrl = $inputFileContent.repoHttpsUrl
+foreach ($inputFilePath in $inputFilePaths) {
+  foreach ($path in $autorestFilesPath) {
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:22]: I would flit these loops so we only read and write autores.md files once and process `inputFilePaths` in the inner loop.

File: eng/scripts/Sdk-Generate.ps1 [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    $fileContent = Get-Content $path
+    $isUpdatedLines = $false
+    $updatedLines = foreach ($line in $fileContent) {
+      $regexForMatchingShaAndPath = "[\/][0-9a-f]{4,40}[\/]$inputFilePath"
+      if ($line -match $regexForMatchingShaAndPath) {
+        $line -replace "https:\/\/[^`"]*$regexForMatchingShaAndPath", "$repoHttpsUrl/blob/$headSha/$inputFilePath"
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:28]: Can we use the same regex both for match and replace?

File: eng/scripts/Sdk-Generate.ps1 [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+$inputFileContent = Get-Content $InputJsonPath | ConvertFrom-Json
+$inputFilePaths = $inputFileContent.changedFiles;
+$inputFilePaths += $inputFileContent.relatedReadmeMdFiles;
+$inputFilePaths = $inputFilePaths | Select-Object -Unique
+
+$changedFilePaths = $inputFilePaths -join "`n";
+Write-Host "List of changed swagger files and related readmes:`n$changedFilePaths`n"
+
+$autorestFilesPath = Get-ChildItem -Path "$RepoRoot/sdk"  -Filter autorest.md -Recurse | Resolve-Path -Relative
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:15]: What is `Resolve-Path -Relative` relative to? Do we need to pass $repoRoot is some way?

**Comment by pakrym** [comment]
nit: casing of the `Sdk-init.ps1` name should be consistent with the `Sdk-Generate`. I would also added some prefix like `Automation-*` so people who see the script would know that this is not something they need to run.

File: swagger_to_sdk_config.json [hunk 4]
```diff
@@ -6,14 +6,19 @@
       "csharp": "",
       "reflect-api-versions": "",
       "license-header": "MICROSOFT_MIT_NO_VERSION",
-      "sdkrel:csharp-sdks-folder": "./sdk",
-      "use": "@microsoft.azure/autorest.csharp@2.3.82"
+      "use": "@microsoft.azure/autorest.csharp@2.3.84",
```
Comment (pakrym) [inline swagger_to_sdk_config.json:9]: Is this change related?

File: eng/scripts/Sdk-Generate.ps1 [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    if ($isUpdatedLines) {
+      $updatedLines | Out-File -FilePath $path
+    }
+  }
+}
+
+Write-Host "Updated autorest.md files for all the changed swaggers. `n"
+
+function Test-PreviousScript() {
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:49]: Is this function mostly for debugging?

File: eng/scripts/Sdk-Generate.ps1 [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    $csprojPath = Get-ChildItem $srcPath -Filter *.csproj -Recurse
+    dotnet pack $csprojPath /p:RunApiCompat=$false
+    $result = Test-PreviousScript
+    if ($result -eq "succeeded") {
+      $artifactsPath = "$RepoRoot/artifacts/packages/Debug/$packageName"
+      $artifacts += Get-ChildItem $artifactsPath -Filter *.nupkg -Recurse | Select-Object -ExpandProperty FullName | Resolve-Path -Relative
+
+      $logFilePath = Join-Path "$srcPath" 'log.txt'
+      if (!(Test-Path $logFilePath)) {
+        New-Item $logFilePath
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:84]: I expect `dotnet build` to create a new log file on it's own. Do we need to make sure that the file is there?

File: eng/scripts/Sdk-Generate.ps1 [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      }
+      dotnet build $csprojPath /t:RunApiCompat /p:TargetFramework=netstandard2.0 /flp:v=m`;LogFile=$logFilePath
+      
+      $result = Test-PreviousScript
+      if ($result -eq "succeeded") {
+        $hasBreakingChange = $false
+      }
+      else {
+        $logFile = Get-Content -Path $logFilePath | select-object -skip 2
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:93]: Are you skipping the msbuild header? you can use ` dotnet build /nologo` to avoid it.

File: eng/scripts/Sdk-Generate.ps1 [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        $breakingChanges = $logFile -join ",`n"
+        $content = "Breaking Changes: $breakingChanges"
+        $hasBreakingChange = $true
+        $result = "succeeded"
+      }
+
+      if (Test-Path $logFilePath) {
+        Remove-Item $logFilePath
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:101]: Any reason we are using the log file instead of capturing the std out? 

`$logs = & dotnet build ... `

File: eng/scripts/Sdk-Generate.ps1 [hunk 9]
```diff
@@ -0,0 +1,144 @@
+param(
+  [string] $InputJsonPath = $GenerateInput,
+  [string] $OutputJsonPath = $GenerateOutput,
```
Comment (pakrym) [inline eng/scripts/Sdk-Generate.ps1:3]: Where `GenerateInput` and `GenerateOutput` are expected to come from?

File: eng/scripts/Sdk-Generate.ps1 [hunk 10]
```diff
@@ -0,0 +1,144 @@
+param(
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:1]: I highly recommend writing scripts as advanced cmdlets, which use the `[CmdletBinding()]` attribute here before `param` and parameters are attributed with `[Parameter()]` and optional properties. See *eng/common/TestResources/New-TestResources.ps1* for examples, or type `help about_advanced_functions` for details.

These have a lot of benefits including pipelining input (if needed), propagating `-Verbose` and `-Debug` through to cmdlets within the script, and much more.

As written, this would work but doesn't really take advantage of PowerShell's features.

File: eng/scripts/Sdk-Generate.ps1 [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  [string] $InputJsonPath = $GenerateInput,
+  [string] $OutputJsonPath = $GenerateOutput,
+  [string] $RepoRoot = "${PSScriptRoot}/../.."
+)
+
+$inputFileContent = Get-Content $InputJsonPath | ConvertFrom-Json
+$inputFilePaths = $inputFileContent.changedFiles;
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:8]: If `changedFiles` is a single object, `$inputFilePaths` will be a string instead of a collection and the `+=` below will fail. Instead, do this:

```powershell
$inputFilePaths = , $inputFileContent.changedFiles
```

The comma is an easy way to force it to be a collection regardless of the rvalue.

File: eng/scripts/Sdk-Generate.ps1 [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+Write-Host "Updating autorest.md files for all the changed swaggers."
+$sdksInfo = @{}
+$headSha = $inputFileContent.headSha
+$repoHttpsUrl = $inputFileContent.repoHttpsUrl
+foreach ($inputFilePath in $inputFilePaths) {
+  foreach ($path in $autorestFilesPath) {
+    $fileContent = Get-Content $path
+    $isUpdatedLines = $false
+    $updatedLines = foreach ($line in $fileContent) {
+      $regexForMatchingShaAndPath = "[\/][0-9a-f]{4,40}[\/]$inputFilePath"
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:26]: `$inputFilePath` is going to have backslashes, periods, and other special regex characters that should be escaped or it won't reliably match. Pass it first to `[System.Text.RegularExpressions.Regex]::Escape([string])`: https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.escape?view=net-5.0

File: eng/scripts/Sdk-Generate.ps1 [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+function Test-PreviousScript() {
+  $result = $null
+  if ($LASTEXITCODE -eq 0) {
+    $result = "succeeded"
+  }
+  else {
+    $result = "failed"
+  }
+  Write-Host "Result: $result"
+  return $result
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:58]: Tip: `return` isn't necessary. All results not captured are written to the pipeline. This is the same as:
```powershell
$result
return
```

`Return` is basically just there for early returns. This isn't wrong, per se, but unnecessary.

Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:58]: Better and easier to just return `$true` or `$false`. Your conditions using this below get much easier:

```powershell
if (test-previousscript) {
  # ...
}
```

File: eng/scripts/Sdk-Generate.ps1 [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  $packageName = Split-Path $sdkPath -Leaf
+  Write-Host "Generating code for " $packageName
+  $srcPath = Join-Path $sdkPath 'src'
+  dotnet msbuild /restore /t:GenerateCode $srcPath
+
+  $artifacts = @()
+  $hasBreakingChange = $null
+  $content = $null
+  $result = Test-PreviousScript
+  if ($result -eq "succeeded") {
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:72]: You really don't even need this. PowerShell uses lazy boolean logic like JavaScript and C/C++. You can run your script and just check $LASTEXITCODE like so:

```
dotnet msbuild ...
if (!$LASTEXITCODE) {
  # continue with successful result...
}
```

There's also the automatic `$?` variable, but that can only be used *immediately* after the command you want to check (PowerShell or native/shell). Even as written, checking `$LASTEXITCODE` too long after you ran a native/shell command rists it being overwritten - it has all the same problems of Win32's `GetLastError()`, i.e. don't want to read it.

File: eng/scripts/Sdk-Generate.ps1 [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+$packages = @()
+foreach ($sdkPath in $sdksInfo.Keys) {
+  $packageName = Split-Path $sdkPath -Leaf
+  Write-Host "Generating code for " $packageName
+  $srcPath = Join-Path $sdkPath 'src'
+  dotnet msbuild /restore /t:GenerateCode $srcPath
+
+  $artifacts = @()
+  $hasBreakingChange = $null
+  $content = $null
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:70]: Tip: this isn't necessary in PowerShell. Variables can be created at any time and are the same as $null if not assigned by the time you read them. They are always available in the scope in which they were declared or assigned. Statements like `if` below don't create new scopes, but scriptblocks (like passed to `foreach-object`) will and then you have to use the `$script:` prefix on a variable to set the parent scope's variables anyway (whether declared or not). See `help about_scopes` for details.

File: eng/scripts/Sdk-Generate.ps1 [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+      $result = Test-PreviousScript
+      if ($result -eq "succeeded") {
+        $hasBreakingChange = $false
+      }
+      else {
+        $logFile = Get-Content -Path $logFilePath | select-object -skip 2
+        $breakingChanges = $logFile -join ",`n"
+        $content = "Breaking Changes: $breakingChanges"
+        $hasBreakingChange = $true
+        $result = "succeeded"
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:97]: I'd stick with `$true` and `$false` rather than inventing your own return codes. It's idiomatic, and easier to maintain. That's what `Test-Path` is returning, for example.

File: eng/scripts/Sdk-Generate.ps1 [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+      if (Test-Path $logFilePath) {
+        Remove-Item $logFilePath
+      }
+    }
+  } 
+  else {
+    Write-Error "Error occurred while generating code for" $packageName "`n"
+  }
+
+  $path = @()
```
Comment (heaths) [inline eng/scripts/Sdk-Generate.ps1:109]:
```powershell
$path = , $sdkPath
```

**Review by heaths — CHANGES_REQUESTED** [review]

File: eng/scripts/Sdk-Generate.ps1 (hunk 3 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: This command returns relative path as generateOutput.json expects. I don't think there is need to pass $repoRoot, currently it returns relative path as expected.

File: swagger_to_sdk_config.json (hunk 4 above)
Comment (ShivangiReja) [reply swagger_to_sdk_config.json]: Reverted. @PhoenixHe-msft did this change. He will create a seperate PR if anythings needs to be updated in configs.

File: eng/scripts/Sdk-Generate.ps1 (hunk 5 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: Removed this function. This function was for testing the script and setting the `result` variable.  This `result` is used in generateOuput.json. Check [this ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=290)example.

File: eng/scripts/Sdk-Generate.ps1 (hunk 7 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: Yes, also I'm skipping these 2 lines
```
Determining projects to restore...
All projects are up-to-date for restore.
```

File: eng/scripts/Sdk-Generate.ps1 (hunk 10 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: Updated! Thanks for the suggestion :)

File: eng/scripts/Sdk-Generate.ps1 (hunk 11 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]:
```
"changedFiles": [
    "specification/cognitiveservices/data-plane/AnomalyDetector/preview/v1.0/AnomalyDetector.json"
  ]
```

`changedFiles` is an array. Even if it will have a single object, `$inputFilePaths` will always be a collection. Please let me know if I'm missing anything.

File: eng/scripts/Sdk-Generate.ps1 (hunk 13 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: Removed this function and used `!$LASTEXITCODE` directly.

Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: Removed this function and used `!$LASTEXITCODE` directly.

File: eng/scripts/Sdk-Generate.ps1 (hunk 15 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: I'm setting these null because if it fails while generating code/creating artifacts [here](https://github.com/Azure/azure-sdk-for-net/blob/c5ef13f487eb3190b8d46be488470b17e1d35aee/eng/scripts/Automation-Sdk-Generate.ps1#L89-L97), it will go to the else block where I'm not setting values for  `$hasBreakingChange` and `$content`, hence it will have previous values(that we don't want to use for the current package). If it fails while generating code/creating artifacts, I want to set null values.

File: eng/scripts/Automation-Sdk-Generate.ps1 [hunk 18]
```diff
@@ -1,6 +1,12 @@
+[CmdletBinding()]
 param(
+  [Parameter(Mandatory = $True)]
   [string] $InputJsonPath = $GenerateInput,
```
Comment (heaths) [inline eng/scripts/Automation-Sdk-Generate.ps1:4]: 1. To @pakrym's question, where are these variables coming from?
2. Putting `Mandatory=$true` on this will require a user still pass a value on the command line. If you want to default it but specify a value is required, so this instead:

```powershell
[Parameter()]
[ValidateNotNullOrEmpty()]
[string] $InputJsonPath = $GenerateInput,
```

This way, if `$GenerateInput` is null or empty (same thing in PowerShell by default in most cases), it'll prompt the user in interactive mode or fail early in non-interactive modes (like during build).

File: eng/scripts/Sdk-Generate.ps1 (hunk 16 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: The values for $result are fixed and it can be either succeeded or failed in generateOuput.json in[ SDK automation](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=271) step. If I set any other value it will fail.

**Review by heaths — APPROVED** [review]

File: eng/scripts/Sdk-Generate.ps1 (hunk 9 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: [GenerateInput ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=107) and [GenerateOutput](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=270) is passed in [SDK Automation step](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5). It is configured in the pipeline.

File: eng/scripts/Sdk-Generate.ps1 (hunk 11 above)
Comment (heaths) [reply eng/scripts/Sdk-Generate.ps1]: PowerShell has some weird behaviors in many places. I'm not sure about `ConvertFrom-Json` specifically, but a lot of other times a single element gets treated as a scalar. I wouldn't change it. Another option is like this:

```powershell
[string[]] $inputFilePaths = $inputFileContent.changedFiles
```

File: eng/scripts/Automation-Sdk-Generate.ps1 (hunk 18 above)
Comment (ShivangiReja) [reply eng/scripts/Automation-Sdk-Generate.ps1:5]: As I mentioned [here](https://github.com/Azure/azure-sdk-for-net/pull/17193#discussion_r533644697) 
>[GenerateInput ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=107) and [GenerateOutput](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=270) are passed in [SDK Automation step](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4254356&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5). It is configured in the pipeline.

Updated!
>[Parameter()]
[ValidateNotNullOrEmpty()]
[string] $InputJsonPath = $GenerateInput

File: eng/scripts/Automation-Sdk-Generate.ps1 [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+  $hasBreakingChange = $null
+  $content = $null
+  $srcPath = Join-Path $sdkPath 'src'
+  dotnet msbuild /restore /t:GenerateCode $srcPath
+  if (!$LASTEXITCODE) {
+    $result = "succeeded"
+    Write-Host "Successfully generated code for" $packageName "`n"
+    
+    $csprojPath = Get-ChildItem $srcPath -Filter *.csproj -Recurse
+    dotnet pack $csprojPath /p:RunApiCompat=$false
```
Comment (pakrym) [inline eng/scripts/Automation-Sdk-Generate.ps1:71]: Sometimes when there are breaking changes the build might fail so you would never get to `dotnet build` line. Not sure how we want to handle that.

File: eng/scripts/Automation-Sdk-Generate.ps1 [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+$autorestFilesPath = Get-ChildItem -Path "$RepoRoot/sdk"  -Filter autorest.md -Recurse | Resolve-Path -Relative
+
+Write-Host "Updating autorest.md files for all the changed swaggers."
+$sdksInfo = @{}
+$headSha = $inputFileContent.headSha
+$repoHttpsUrl = $inputFileContent.repoHttpsUrl
+foreach ($path in $autorestFilesPath) {
+  foreach ($inputFilePath in $inputFilePaths) {
+    $fileContent = Get-Content $path
```
Comment (pakrym) [inline eng/scripts/Automation-Sdk-Generate.ps1:30]: You can move this into the outer scope and well as writing.

File: eng/scripts/Automation-Sdk-Generate.ps1 (hunk 18 above)
Comment (pakrym) [reply eng/scripts/Automation-Sdk-Generate.ps1:5]: I don't think you need the ` = $GenerateInput` part. You would get the parameter passed into the script without it.

File: eng/scripts/Automation-Sdk-Generate.ps1 [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+      if (!$LASTEXITCODE) {
+        $result = "succeeded"
+        $hasBreakingChange = $false
+      }
+      else {
+        $logFile = $logs | select-object -skip 2
+        $breakingChanges = $logFile -join ",`n"
+        $content = "Breaking Changes: $breakingChanges"
+        $hasBreakingChange = $true
+        $result = "succeeded"
```
Comment (pakrym) [inline eng/scripts/Automation-Sdk-Generate.ps1:87]: You don't need this line, result is already set here.

File: eng/scripts/Automation-Sdk-Generate.ps1 [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+    $csprojPath = Get-ChildItem $srcPath -Filter *.csproj -Recurse
+    dotnet pack $csprojPath /p:RunApiCompat=$false
+    if (!$LASTEXITCODE) {
+      $result = "succeeded"
+      $artifactsPath = "$RepoRoot/artifacts/packages/Debug/$packageName"
+      $artifacts += Get-ChildItem $artifactsPath -Filter *.nupkg -Recurse | Select-Object -ExpandProperty FullName | Resolve-Path -Relative
+
+      $logs = & dotnet build /nologo $csprojPath /t:RunApiCompat /p:TargetFramework=netstandard2.0 /flp:v=m`;
+      if (!$LASTEXITCODE) {
+        $result = "succeeded"
```
Comment (pakrym) [inline eng/scripts/Automation-Sdk-Generate.ps1:79]: You don't need this line, result is already set here.

**Review by pakrym — APPROVED** [review]

File: eng/scripts/Automation-Sdk-Generate.ps1 (hunk 19 above)
Comment (ShivangiReja) [reply eng/scripts/Automation-Sdk-Generate.ps1:71]: If the build will fail it go to the else block [here](https://github.com/Azure/azure-sdk-for-net/blob/6913ed4d48856a206f98d2b728750a3e9ffe2673/eng/scripts/Automation-Sdk-Generate.ps1#L91) and It will still create a PR. Then we will have manually fix the issues.

Check Digital twin example [here](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4264729&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=193), it failed while creating an artifacts but [here ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4264729&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=341)it created a PR

File: eng/scripts/Automation-Sdk-Generate.ps1 (hunk 18 above)
Comment (ShivangiReja) [reply eng/scripts/Automation-Sdk-Generate.ps1]: Yes, you're right. 

Updated!

File: eng/scripts/Automation-Sdk-Generate.ps1 (hunk 19 above)
Comment (weshaggard) [reply eng/scripts/Automation-Sdk-Generate.ps1:71]: Depending on the changes it the build might fail but it should be mostly limited to the individual project and not span to other projects because we don't use project references.

File: eng/scripts/Sdk-Generate.ps1 (hunk 8 above)
Comment (weshaggard) [reply eng/scripts/Sdk-Generate.ps1]: In general I like the log files to keep things clean and also to ensure we get useful out put in the step when running without having to also dump the content of the captured string after the command so we see forward progress, if we are trying to debug what might be going wrong in a build.

File: eng/scripts/Sdk-Generate.ps1 (hunk 15 above)
Comment (weshaggard) [reply eng/scripts/Sdk-Generate.ps1]: Depending on the strict mode (https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/set-strictmode?view=powershell-7.1) you might actually need to initialize these values.  I've personally started setting Strict mode version 3 in my scripts to help catch stupid issues with uninitialized variables where I had a typo or something.

File: eng/scripts/Sdk-Generate.ps1 (hunk 17 above)
Comment (weshaggard) [reply eng/scripts/Sdk-Generate.ps1]: I have a love hate relation ship with the comma operator :)

File: eng/scripts/Automation-Sdk-Generate.ps1 [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+  $artifacts = @()
+  $hasBreakingChange = $null
+  $content = $null
+  $srcPath = Join-Path $sdkPath 'src'
+  dotnet msbuild /restore /t:GenerateCode $srcPath
+  if (!$LASTEXITCODE) {
+    $result = "succeeded"
+    Write-Host "Successfully generated code for" $packageName "`n"
+    
+    $csprojPath = Get-ChildItem $srcPath -Filter *.csproj -Recurse
```
Comment (weshaggard) [inline eng/scripts/Automation-Sdk-Generate.ps1:70]: We are making an assumption that there is only one csproj found. If this ever returns more then one the line below it will fail in an unknown way. It might be worth checking that we only have one path.

Actually we could even consider just calling dotnet pack passing in the $srcPath as I think it will also find the project or give you an error about multiple projects if it finds more than one.

File: swagger_to_sdk_config_track2.json [hunk 24]
```diff
@@ -0,0 +1,17 @@
+{
+  "$schema": "https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/sdkautomation/SwaggerToSdkConfigSchema.json",
+  "advancedOptions": {
+    "createSdkPullRequests": true,
+    "generationCallMode": "one-for-all-configs"
```
Comment (weshaggard) [inline swagger_to_sdk_config_track2.json:5]: Is this required? I thought this was the default.

Comment (weshaggard) [reply swagger_to_sdk_config_track2.json:5]: That goes for all the advancedOptions. I think we can remove them and only have the init and generate script configurations.

Comment (ShivangiReja) [reply swagger_to_sdk_config_track2.json:5]: I'm not sure about it. @PhoenixHe-msft will update the configs after we merge this PR.

File: swagger_to_sdk_config_track2.json [hunk 25]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (weshaggard) [inline swagger_to_sdk_config_track2.json:1]: Can we move this from the root of the repo to under "eng"?

Comment (weshaggard) [reply swagger_to_sdk_config_track2.json:1]: Also it would be nice to not use "track2" as the name of the configuration as that is more of a code name.  Given we are putting it in a unique path lets just call it "swagger_to_sdk_config.json"

File: eng/scripts/Automation-Sdk-Generate.ps1 (hunk 23 above)
Comment (ShivangiReja) [reply eng/scripts/Automation-Sdk-Generate.ps1]: Updated!

File: eng/swagger_to_sdk_config_track.json [hunk 26]
```diff
@@ -1,17 +1,13 @@
 {
   "$schema": "https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/sdkautomation/SwaggerToSdkConfigSchema.json",
-  "advancedOptions": {
-    "createSdkPullRequests": true,
-    "generationCallMode": "one-for-all-configs"
-  },
   "generateOptions": {
     "generateScript": {
-      "path": "pwsh ./eng/scripts/Automation-Sdk-Generate.ps1"
+      "path": "pwsh ./scripts/Automation-Sdk-Generate.ps1"
```
Comment (weshaggard) [inline eng/swagger_to_sdk_config_track.json]: It is probably worth testing to ensure this path is supposed to be relative to the json file vs the root of the repo?

**Review by weshaggard — APPROVED** [review]
The initial set of changes look good. Lets get this merged and start getting some services opted into using it.

Thanks for all the work!

File: eng/swagger_to_sdk_config_track.json (hunk 26 above)
Comment (ShivangiReja) [reply eng/swagger_to_sdk_config_track.json]: Thanks for noticing. It should be the root of the repo.

Path updated!

File: eng/scripts/Sdk-Generate.ps1 (hunk 8 above)
Comment (ShivangiReja) [reply eng/scripts/Sdk-Generate.ps1]: Reverted the logic back to creating log file. As @weshaggard said and I also observed that the output of the log file is nice and clean whereas std out is capturing lots of useless strings that we don't want as a part of breaking changes.

For reference:
* [Here ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4273481&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=3045)(Line 3045-3059) are the breaking changes captured by the log file.
* [Here ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4273398&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=2988)(Line 2988-3024)  are the breaking changes captured by the std out.

File: eng/scripts/Automation-Sdk-Init.ps1 [hunk 27]
```diff
@@ -2,13 +2,13 @@
 [string] $dotnetInstallScriptVersion = "v1"
 
 function GetDotNetInstallScript() {
-    $installScript = Join-Path $RepoRoot 'dotnet-install.ps1'
+    $installScript = Join-Path $RepoRoot 'dotnet-install.sh'
```
Comment (heaths) [inline eng/scripts/Automation-Sdk-Init.ps1:5]: Presumably, this will only work on non-Windows machines. Why the change? The ps1 file will also run in the current session instead of shelling out to a separate process, which can often lead to issues with parameter passing (especially when quotes are involved).

Comment (ShivangiReja) [reply eng/scripts/Automation-Sdk-Init.ps1:5]: Yes, we will run this only on linux. Previous changes were giving me [these](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4266624&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=93) errors because I wrote the commands for Windows machines. With the current changes everything is [up and running](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4273481&view=logs&j=7f026a02-f755-5ce4-425c-2d5df22c9d55&t=20b436ef-0412-5b55-7996-5294882c57b5&l=86).
